File: lib/analyzer/scorer.ts
// TF-IDF cosine similarity — hand-rolled so we can run it in a Vercel function
// without a Python sklearn dependency. Results are directionally correct for MVP
// scoring; see ROADMAP for a note on parity with sklearn's implementation.

const STOPWORDS = new Set([
  'a','an','the','and','or','but','in','on','at','to','for','of','with','by',
  'from','up','about','into','through','during','is','are','was','were','be',
  'been','being','have','has','had','do','does','did','will','would','could',
  'should','may','might','that','this','these','those','it','its','as','not',
  'no','so','if','then','than','can','our','we','you','your','my','their','i',
  'me','him','her','us','them','who','what','when','where','how','which','all',
  'each','both','few','more','most','other','some','such','only','own','same',
])

export function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, ' ')
    .split(/\s+/)
    .filter(t => t.length > 2 && !STOPWORDS.has(t))
}

function buildTF(tokens: string[]): Map<string, number> {
  const freq = new Map<string, number>()
  for (const t of tokens) freq.set(t, (freq.get(t) ?? 0) + 1)
  // Normalize by document length
  freq.forEach((v, k) => freq.set(k, v / tokens.length))
  return freq
}

function cosine(a: number[], b: number[]): number {
  let dot = 0, normA = 0, normB = 0
  for (let i = 0; i < a.length; i++) {
    dot   += a[i] * b[i]
    normA += a[i] * a[i]
    normB += b[i] * b[i]
  }
  if (normA === 0 || normB === 0) return 0
  // Clamp to [0,1] to handle floating-point noise
  return Math.min(1, dot / (Math.sqrt(normA) * Math.sqrt(normB)))
}

export function tfidfScore(docA: string, docB: string): number {
  const tokensA = tokenize(docA)
  const tokensB = tokenize(docB)
  const tfA = buildTF(tokensA)
  const tfB = buildTF(tokensB)

  const vocab = new Set([...tfA.keys(), ...tfB.keys()])

  // IDF over a two-document corpus: log(2 / df)
  const idf = new Map<string, number>()
  for (const term of vocab) {
    const df = (tfA.has(term) ? 1 : 0) + (tfB.has(term) ? 1 : 0)
    idf.set(term, Math.log(2 / df))
  }

  const vecA: number[] = []
  const vecB: number[] = []
  for (const term of vocab) {
    const w = idf.get(term) ?? 0
    vecA.push((tfA.get(term) ?? 0) * w)
    vecB.push((tfB.get(term) ?? 0) * w)
  }

  return cosine(vecA, vecB)
}

export function scoreSections(
  sections: Record<string, string>,
  referenceText: string,
): Array<{ section: string; score: number }> {
  return Object.entries(sections)
    .filter(([, content]) => content.trim().length > 20)
    .map(([section, content]) => ({
      section,
      score: Math.round(tfidfScore(content, referenceText) * 100),
    }))
    .sort((a, b) => b.score - a.score)
}

/**
 * Fraction of JD requirements TF-IDF mass that is covered by resume tokens,
 * in [0, 1]. Returns 0.5 (neutral) when inputs are empty.
 *
 * Uses the same hand-rolled TF-IDF as tfidfScore but measures coverage
 * (how much of the JD's term mass does the resume address?) rather than
 * symmetric cosine similarity.
 */
export function requirementsCoverage(
  resumeText: string,
  requirementsText: string,
): number {
  if (!resumeText.trim() || !requirementsText.trim()) return 0.5

  const reqTokens = tokenize(requirementsText)
  const resumeTokens = tokenize(resumeText)
  const resumeSet = new Set(resumeTokens)

  // Build TF-IDF weights for requirements text only
  const reqTF = buildTF(reqTokens)

  // IDF computed over a two-doc corpus (requirements + resume) to match sklearn semantics
  const allVocab = new Set([...reqTokens, ...resumeTokens])
  const resumeTF = buildTF(resumeTokens)

  let totalMass = 0
  let coveredMass = 0

  for (const term of allVocab) {
    const df = (reqTF.has(term) ? 1 : 0) + (resumeTF.has(term) ? 1 : 0)
    const idf = Math.log(2 / df)
    const reqWeight = (reqTF.get(term) ?? 0) * idf

    totalMass += reqWeight
    if (resumeSet.has(term)) {
      coveredMass += reqWeight
    }
  }

  if (totalMass === 0) return 0.5
  return Math.min(1, coveredMass / totalMass)
}

export function keywordGap(resumeText: string, jdText: string): string[] {
  const resumeTokens = new Set(tokenize(resumeText))
  const jdTokens = tokenize(jdText)

  // Rank JD terms by raw frequency, then filter out those already in the resume
  const jdFreq = new Map<string, number>()
  for (const t of jdTokens) jdFreq.set(t, (jdFreq.get(t) ?? 0) + 1)

  return Array.from(jdFreq.entries())
    .filter(([term]) => !resumeTokens.has(term) && term.length > 3)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 30)
    .map(([term]) => term)
}

File: api/app/services/scorer.py
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np


def tfidf_score(doc_a: str, doc_b: str) -> float:
    """Cosine similarity between two documents using sklearn's TF-IDF.

    Using sklearn here (vs the hand-rolled TypeScript version) gives us the
    full sublinear TF smoothing and IDF calibrated across a larger corpus
    of stop-word-filtered terms, which produces better-calibrated scores.
    """
    if not doc_a.strip() or not doc_b.strip():
        return 0.0

    vectorizer = TfidfVectorizer(
        lowercase=True,
        strip_accents="unicode",
        stop_words="english",
        ngram_range=(1, 1),
        sublinear_tf=True,
        min_df=1,
    )
    try:
        tfidf = vectorizer.fit_transform([doc_a, doc_b])
        score = float(cosine_similarity(tfidf[0:1], tfidf[1:2])[0][0])
    except ValueError:
        # sklearn raises this when stop-word filtering leaves an empty vocabulary
        return 0.0
    return max(0.0, min(1.0, score))


def score_sections(sections: dict[str, str], jd_text: str) -> list[dict]:
    """Score each detected resume section against a reference text (usually requirements)."""
    results = []
    for section, content in sections.items():
        if len(content.strip()) < 20:
            continue
        score = round(tfidf_score(content, jd_text) * 100)
        results.append({"section": section, "score": score})
    return sorted(results, key=lambda x: x["score"], reverse=True)


def requirements_coverage(resume_text: str, requirements_text: str) -> float:
    """Fraction of JD requirements TF-IDF mass that is covered by resume tokens.

    Unlike symmetric cosine similarity, this metric asks: "how much of what the
    job demands does the resume address?" — a better proxy for recruiter fit.

    Returns a value in [0, 1]. Returns 0.5 (neutral) on empty inputs so the
    composite score is not dragged down when requirements can't be extracted.
    """
    if not requirements_text.strip() or not resume_text.strip():
        return 0.5

    vectorizer = TfidfVectorizer(
        lowercase=True,
        strip_accents="unicode",
        stop_words="english",
        ngram_range=(1, 1),
        sublinear_tf=True,
        min_df=1,
    )
    try:
        vectorizer.fit([requirements_text])
        jd_vec: np.ndarray = vectorizer.transform([requirements_text]).toarray()[0]
        resume_vec: np.ndarray = vectorizer.transform([resume_text]).toarray()[0]
    except ValueError:
        return 0.5

    total_mass = float(jd_vec.sum())
    if total_mass == 0:
        return 0.5

    # Sum JD term weights where the resume also has a non-zero score for that term
    covered_mass = float(np.sum(jd_vec[resume_vec > 0]))
    return min(1.0, covered_mass / total_mass)


def keyword_gap(resume_text: str, jd_text: str) -> list[str]:
    """Return high-frequency JD terms absent from the resume, ranked by importance.

    We fit TF-IDF on the JD alone to rank its terms, then filter out any that
    appear (case-insensitively) in the resume text.
    """
    if not jd_text.strip():
        return []

    vectorizer = TfidfVectorizer(
        lowercase=True,
        stop_words="english",
        ngram_range=(1, 1),
        min_df=1,
    )
    try:
        vectorizer.fit([jd_text])
        terms = vectorizer.get_feature_names_out()
        scores = vectorizer.transform([jd_text]).toarray()[0]
    except ValueError:
        return []

    resume_lower = resume_text.lower()
    missing = [
        term for term, score in sorted(zip(terms, scores), key=lambda x: -x[1])
        if len(term) > 3 and term not in resume_lower
    ]
    return missing[:30]
